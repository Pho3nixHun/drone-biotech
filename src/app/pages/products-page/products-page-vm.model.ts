import { FrameVM } from '@components/frame/frame-vm';
import { ProductItemXVM } from '../landing-page/landing-page-vm.model';

interface WithProductItemXVMs {
    productItemVMs: ProductItemXVM[] | null;
}

interface ProductListFrame extends FrameVM, WithProductItemXVMs {}

export interface ProductsPageVM {
    productListFrame: ProductListFrame;
}
