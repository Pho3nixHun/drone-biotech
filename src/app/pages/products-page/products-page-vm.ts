import { FrameVM } from '@components/frame/frame-vm';
import { ProductItemComponentVM } from '@components/product-item/product-item.component.vm';

export interface ProductsPageVM {
  frameVM: FrameVM;
  productItems: ProductItemComponentVM[];
}
