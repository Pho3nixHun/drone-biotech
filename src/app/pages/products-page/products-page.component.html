@if (vm(); as vm) {
    <app-page-layout>
        @if (vm.productListFrame.productItemVMs) {
            <app-frame>
                <h2>{{ t(vm.productListFrame.titleKey) }}</h2>
                <app-product-list>
                    @for (
                        item of vm.productListFrame.productItemVMs;
                        track $index
                    ) {
                        <app-product-item [vm]="item" />
                    }
                </app-product-list>
            </app-frame>
        }
    </app-page-layout>
}
