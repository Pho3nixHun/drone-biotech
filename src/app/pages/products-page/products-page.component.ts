import { Component, inject } from '@angular/core';
import { FrameComponent } from '@components/frame/frame.component';
import { ProductItemComponent } from '@components/product-item/product-item.component';
import { ProductListComponent } from '@components/product-list/product-list.component';
import { ProductsPageService } from './products-page.service';
import { PageSectionComponent } from '../../shared/components/page-section/page-section.component';

/**
 * ProductsPageComponent
 *
 * Type: Container
 *
 * Scope:
 * - Conditionally renders a section with a product or with a list of products based on the URL.
 * - Passes relevant data.
 *
 * Out-of-Scope:
 * - Does not handle the internal logic or styling of the `app-product-list`, <app-product-item> components.
 * - Not responsible for the detailed presentation logic.
 *
 * Purpose (optional):
 * - To serve as a smart container component that integrates business logic, including data fetching and presentation, to create a cohesive user interface.
 */
@Component({
    selector: 'app-products-page',
    imports: [
        FrameComponent,
        ProductItemComponent,
        ProductListComponent,
        PageSectionComponent,
    ],
    templateUrl: './products-page.component.html'
})
export class ProductsPageComponent {
    private readonly productsPageService = inject(ProductsPageService);
    protected readonly vm = this.productsPageService.getVM();
}
