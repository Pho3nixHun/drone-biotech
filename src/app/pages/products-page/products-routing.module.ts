import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';

export const routes: Routes = [
    {
        path: '',
        loadComponent: () =>
            import('../products-page/products-page.component').then(
                (m) => m.ProductsPageComponent
            ),
    },
    {
        path: ':id',
        loadComponent: () =>
            import('../products-page/products-page.component').then(
                (m) => m.ProductsPageComponent
            ),
    },
];

@NgModule({
    imports: [RouterModule.forChild(routes)],
    exports: [RouterModule],
})
export class ProductsRoutingModule {}
