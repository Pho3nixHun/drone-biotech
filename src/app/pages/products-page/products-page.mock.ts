import { ProductsPageVM } from './products-page-vm.model';

export const enProductsPageMock = {
    title0: 'title0',
    productAltText0: 'title0',
    productDescription0: 'title1',
    productAltText1: 'title0',
    productDescription1: 'title1',
    productAltText2: 'title0',
    productDescription2: 'title1',
};

export const productsPageVMWithoutProductItem: ProductsPageVM = {
    productListFrame: {
        titleKey: 'title0',
        productItemVMs: [],
    },
};

export const productsPageVMWithOneProductItem: ProductsPageVM = {
    productListFrame: {
        titleKey: 'title0',
        productItemVMs: [
            {
                routerLink: '1',
                id: '1',
                altTextKey: 'productAltText0',
                titleKey: 'productAltText0',
                descriptionKey: 'productDescription0',
                imageSrc: 'assets/lepke.jpg',
            },
        ],
    },
};

export const productsPageVMWithFiveProductItem: ProductsPageVM = {
    productListFrame: {
        titleKey: 'title0',
        productItemVMs: [
            {
                routerLink: '1',
                id: '1',
                altTextKey: 'productAltText0',
                titleKey: 'productAltText0',
                descriptionKey: 'productDescription0',
                imageSrc: 'assets/lepke.jpg',
            },
            {
                routerLink: '2',
                id: '2',
                altTextKey: 'productAltText1',
                titleKey: 'productAltText1',
                descriptionKey: 'productDescription1',
                imageSrc: 'assets/lepke.jpg',
            },
            {
                routerLink: '3',
                id: '3',
                altTextKey: 'productAltText2',
                titleKey: 'productAltText2',
                descriptionKey: 'productDescription2',
                imageSrc: 'assets/lepke.jpg',
            },
            {
                routerLink: '2',
                id: '2',
                altTextKey: 'productAltText1',
                titleKey: 'productAltText1',
                descriptionKey: 'productDescription1',
                imageSrc: 'assets/lepke.jpg',
            },
            {
                routerLink: '3',
                id: '3',
                altTextKey: 'productAltText2',
                titleKey: 'productAltText2',
                descriptionKey: 'productDescription2',
                imageSrc: 'assets/lepke.jpg',
            },
        ],
    },
};

export const productsPageVMWithRoutes: ProductsPageVM = {
    productListFrame: {
        titleKey: 'title0',
        productItemVMs: [
            {
                routerLink: '1',
                id: '1',
                altTextKey: 'productAltText0',
                titleKey: 'productAltText0',
                descriptionKey: 'productDescription0',
                imageSrc: 'assets/lepke.jpg',
            },
            {
                routerLink: '2',
                id: '2',
                altTextKey: 'productAltText1',
                titleKey: 'productAltText1',
                descriptionKey: 'productDescription1',
                imageSrc: 'assets/lepke.jpg',
            },
            {
                routerLink: '3',
                id: '3',
                altTextKey: 'productAltText2',
                titleKey: 'productAltText2',
                descriptionKey: 'productDescription2',
                imageSrc: 'assets/lepke.jpg',
            },
        ],
    },
};

export const productsPageVMMock: ProductsPageVM = {
    productListFrame: {
        titleKey: 'title0',
        productItemVMs: [
            {
                routerLink: '1',
                id: '1',
                altTextKey: 'productAltText0',
                titleKey: 'productAltText0',
                descriptionKey: 'productDescription0',
                imageSrc: 'assets/lepke.jpg',
            },
            {
                routerLink: '2',
                id: '2',
                altTextKey: 'productAltText1',
                titleKey: 'productAltText1',
                descriptionKey: 'productDescription1',
                imageSrc: 'assets/lepke.jpg',
            },
            {
                routerLink: '3',
                id: '3',
                altTextKey: 'productAltText2',
                titleKey: 'productAltText2',
                descriptionKey: 'productDescription2',
                imageSrc: 'assets/lepke.jpg',
            },
        ],
    },
};

export const productsPageVMDefault: ProductsPageVM = {
    productListFrame: {
        titleKey: 'LandingPage.frames.0.title',
        productItemVMs: [
            {
                routerLink: '1',
                id: '1',
                altTextKey: 'LandingPage.products.0.title',
                titleKey: 'LandingPage.products.0.title',
                descriptionKey: 'LandingPage.products.0.description',
                imageSrc: 'assets/lepke.jpg',
            },
            {
                routerLink: '2',
                id: '2',
                altTextKey: 'LandingPage.products.1.title',
                titleKey: 'LandingPage.products.1.title',
                descriptionKey: 'LandingPage.products.1.description',
                imageSrc: 'assets/lepke.jpg',
            },
            {
                routerLink: '3',
                id: '3',
                altTextKey: 'LandingPage.products.2.title',
                titleKey: 'LandingPage.products.2.title',
                descriptionKey: 'LandingPage.products.2.description',
                imageSrc: 'assets/lepke.jpg',
            },
        ],
    },
};
