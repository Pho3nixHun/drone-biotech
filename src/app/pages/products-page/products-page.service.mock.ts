import { Injectable, signal } from '@angular/core';
import { ProductsPageVM } from './products-page-vm.model';

const getVMSignal = signal<ProductsPageVM | undefined>(undefined);

@Injectable({
  providedIn: 'root',
})
class ProductsPageMockService {
  public getVM = () => getVMSignal;
}

export const provideProductsPageMockService = () => ({
    provide: 'ProductsPageService',
    useClass: ProductsPageMockService,
});

export const updateGetVMSignal = getVMSignal.set.bind(getVMSignal);
