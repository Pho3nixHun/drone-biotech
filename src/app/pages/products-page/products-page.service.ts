import { Injectable, signal, Signal } from '@angular/core';
import { ProductsPageVM } from './products-page-vm.model';
import { productsPageVM } from './products-page.mock';

@Injectable({
  providedIn: 'root',
})
export class ProductsPageService {
  public getVM(): Signal<ProductsPageVM | null> {
    return signal(productsPageVM);
  }
}
