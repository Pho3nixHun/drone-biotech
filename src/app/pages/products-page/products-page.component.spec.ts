import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ProductsPageComponent } from './products-page.component';
import { getTranslocoModule } from 'transloco-testing.module';
import { ActivatedRoute } from '@angular/router';
import { enProductsPageMock, productsPageVMMock } from './products-page.mock';
import {
    ACTIVATED_ROUTE_TOKEN,
    provideActivatedRoute,
    provideProductsPageMockService,
    updateGetVMSignal,
    updateRouteSignal,
} from './products-page.service.mock';

describe('ProductsPageComponent', () => {
    let fixture: ComponentFixture<ProductsPageComponent>;
    let compiled: HTMLElement;

    beforeEach(() => {
        TestBed.configureTestingModule({
            imports: [
                ProductsPageComponent,
                getTranslocoModule({
                    langs: { enProductsPageMock },
                    translocoConfig: {
                        availableLangs: ['en'],
                        defaultLang: 'en',
                    },
                }),
            ],
            providers: [
                provideProductsPageMockService(),
                provideActivatedRoute(),
                { provide: ActivatedRoute, useExisting: ACTIVATED_ROUTE_TOKEN },
            ],
        }).compileComponents();

        updateGetVMSignal(undefined);
        TestBed.inject(ActivatedRoute);
        fixture = TestBed.createComponent(ProductsPageComponent);
        fixture.detectChanges();
        compiled = fixture.debugElement.nativeElement;
    });

    //Snapshot test
    it('should not render the template when there is no VM provided', () => {
        //Arrange
        // No need to arrange

        //Act
        //No need to act

        //Assert
        expect(compiled).toMatchSnapshot();
    });
    //Snapshot test
    it('should render the template when the VM is provided', () => {
        //Arrange
        updateGetVMSignal(productsPageVMMock);

        //Act
        fixture.detectChanges();

        //Assert
        expect(compiled).toMatchSnapshot();
    });

    //Snapshot test
    it('should render the correct <app-product-item> based on the current URL', () => {
        //Arrange
        productsPageVMMock.productListFrame.productItemVMs.forEach(
            (element) => {
                updateRouteSignal(element.routerLink);
                updateGetVMSignal(productsPageVMMock);

                //Act
                fixture.detectChanges();

                //Assert
                expect(compiled).toMatchSnapshot();
            }
        );
    });

    //Snapshot test
    it('should not render <app-product-item> if the id is not correct', async () => {
        //Arrange
        updateRouteSignal('route');
        updateGetVMSignal(productsPageVMMock);

        //Act
        fixture.detectChanges();

        //Assert
        expect(compiled).toMatchSnapshot();
    });
});
