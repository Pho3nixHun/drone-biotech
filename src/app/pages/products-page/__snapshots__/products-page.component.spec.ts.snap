// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsPageComponent should not render the template when there is no VM provided 1`] = `
<div
  id="root0"
  ng-version="19.2.9"
>
  <!--container-->
</div>
`;

exports[`ProductsPageComponent should render 1 <app-product-item> if 1 items are provided 1`] = `
<div
  id="root2"
  ng-version="19.2.9"
>
  <app-page-layout>
    <section
      class="flex flex-col gap-y-10 mx-5 sm:mx-20 lg:mx-32"
    >
      <app-frame>
        <section
          class="flex flex-col gap-y-10 py-10"
        >
          <h2>
            title0
          </h2>
          <app-product-list>
            <div
              class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-8 mt-10 transition-all"
            >
              <app-product-item
                class="h-full"
                ng-reflect-vm="[object Object]"
              >
                <a
                  class="cursor-pointer block p-4 w-full h-full rounded-md shadow-md hover:shadow-xl transition-all"
                >
                  <img
                    alt="LandingPage.products.0.title"
                    class="w-auto h-auto object-cover rounded-md"
                    src="assets/lepke.jpg"
                  />
                  <h3
                    class="font-bold text-xl sm:text-2xl mt-4"
                  >
                    LandingPage.products.0.title
                  </h3>
                  <p
                    class="mt-4"
                  >
                    LandingPage.products.0.description
                  </p>
                </a>
                <!--container-->
              </app-product-item>
              <app-product-item
                class="h-full"
                ng-reflect-vm="[object Object]"
              >
                <a
                  class="cursor-pointer block p-4 w-full h-full rounded-md shadow-md hover:shadow-xl transition-all"
                >
                  <img
                    alt="LandingPage.products.1.title"
                    class="w-auto h-auto object-cover rounded-md"
                    src="assets/lepke.jpg"
                  />
                  <h3
                    class="font-bold text-xl sm:text-2xl mt-4"
                  >
                    LandingPage.products.1.title
                  </h3>
                  <p
                    class="mt-4"
                  >
                    LandingPage.products.1.description
                  </p>
                </a>
                <!--container-->
              </app-product-item>
              <app-product-item
                class="h-full"
                ng-reflect-vm="[object Object]"
              >
                <a
                  class="cursor-pointer block p-4 w-full h-full rounded-md shadow-md hover:shadow-xl transition-all"
                >
                  <img
                    alt="LandingPage.products.2.title"
                    class="w-auto h-auto object-cover rounded-md"
                    src="assets/lepke.jpg"
                  />
                  <h3
                    class="font-bold text-xl sm:text-2xl mt-4"
                  >
                    LandingPage.products.2.title
                  </h3>
                  <p
                    class="mt-4"
                  >
                    LandingPage.products.2.description
                  </p>
                </a>
                <!--container-->
              </app-product-item>
              <!--container-->
            </div>
          </app-product-list>
        </section>
      </app-frame>
      <!--container-->
    </section>
  </app-page-layout>
  <!--container-->
  <!--container-->
</div>
`;

exports[`ProductsPageComponent should render 5 <app-product-item> if 5 items are provided 1`] = `
<div
  id="root3"
  ng-version="19.2.9"
>
  <app-page-layout>
    <section
      class="flex flex-col gap-y-10 mx-5 sm:mx-20 lg:mx-32"
    >
      <app-frame>
        <section
          class="flex flex-col gap-y-10 py-10"
        >
          <h2>
            title0
          </h2>
          <app-product-list>
            <div
              class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-8 mt-10 transition-all"
            >
              <app-product-item
                class="h-full"
                ng-reflect-vm="[object Object]"
              >
                <a
                  class="cursor-pointer block p-4 w-full h-full rounded-md shadow-md hover:shadow-xl transition-all"
                >
                  <img
                    alt="LandingPage.products.0.title"
                    class="w-auto h-auto object-cover rounded-md"
                    src="assets/lepke.jpg"
                  />
                  <h3
                    class="font-bold text-xl sm:text-2xl mt-4"
                  >
                    LandingPage.products.0.title
                  </h3>
                  <p
                    class="mt-4"
                  >
                    LandingPage.products.0.description
                  </p>
                </a>
                <!--container-->
              </app-product-item>
              <app-product-item
                class="h-full"
                ng-reflect-vm="[object Object]"
              >
                <a
                  class="cursor-pointer block p-4 w-full h-full rounded-md shadow-md hover:shadow-xl transition-all"
                >
                  <img
                    alt="LandingPage.products.1.title"
                    class="w-auto h-auto object-cover rounded-md"
                    src="assets/lepke.jpg"
                  />
                  <h3
                    class="font-bold text-xl sm:text-2xl mt-4"
                  >
                    LandingPage.products.1.title
                  </h3>
                  <p
                    class="mt-4"
                  >
                    LandingPage.products.1.description
                  </p>
                </a>
                <!--container-->
              </app-product-item>
              <app-product-item
                class="h-full"
                ng-reflect-vm="[object Object]"
              >
                <a
                  class="cursor-pointer block p-4 w-full h-full rounded-md shadow-md hover:shadow-xl transition-all"
                >
                  <img
                    alt="LandingPage.products.2.title"
                    class="w-auto h-auto object-cover rounded-md"
                    src="assets/lepke.jpg"
                  />
                  <h3
                    class="font-bold text-xl sm:text-2xl mt-4"
                  >
                    LandingPage.products.2.title
                  </h3>
                  <p
                    class="mt-4"
                  >
                    LandingPage.products.2.description
                  </p>
                </a>
                <!--container-->
              </app-product-item>
              <!--container-->
            </div>
          </app-product-list>
        </section>
      </app-frame>
      <!--container-->
    </section>
  </app-page-layout>
  <!--container-->
  <!--container-->
</div>
`;

exports[`ProductsPageComponent should render the template when the VM is provided 1`] = `
<div
  id="root1"
  ng-version="19.2.9"
>
  <app-page-layout>
    <section
      class="flex flex-col gap-y-10 mx-5 sm:mx-20 lg:mx-32"
    >
      <app-frame>
        <section
          class="flex flex-col gap-y-10 py-10"
        >
          <h2>
            title0
          </h2>
          <app-product-list>
            <div
              class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-8 mt-10 transition-all"
            >
              <app-product-item
                class="h-full"
                ng-reflect-vm="[object Object]"
              >
                <a
                  class="cursor-pointer block p-4 w-full h-full rounded-md shadow-md hover:shadow-xl transition-all"
                >
                  <img
                    alt="LandingPage.products.0.title"
                    class="w-auto h-auto object-cover rounded-md"
                    src="assets/lepke.jpg"
                  />
                  <h3
                    class="font-bold text-xl sm:text-2xl mt-4"
                  >
                    LandingPage.products.0.title
                  </h3>
                  <p
                    class="mt-4"
                  >
                    LandingPage.products.0.description
                  </p>
                </a>
                <!--container-->
              </app-product-item>
              <app-product-item
                class="h-full"
                ng-reflect-vm="[object Object]"
              >
                <a
                  class="cursor-pointer block p-4 w-full h-full rounded-md shadow-md hover:shadow-xl transition-all"
                >
                  <img
                    alt="LandingPage.products.1.title"
                    class="w-auto h-auto object-cover rounded-md"
                    src="assets/lepke.jpg"
                  />
                  <h3
                    class="font-bold text-xl sm:text-2xl mt-4"
                  >
                    LandingPage.products.1.title
                  </h3>
                  <p
                    class="mt-4"
                  >
                    LandingPage.products.1.description
                  </p>
                </a>
                <!--container-->
              </app-product-item>
              <app-product-item
                class="h-full"
                ng-reflect-vm="[object Object]"
              >
                <a
                  class="cursor-pointer block p-4 w-full h-full rounded-md shadow-md hover:shadow-xl transition-all"
                >
                  <img
                    alt="LandingPage.products.2.title"
                    class="w-auto h-auto object-cover rounded-md"
                    src="assets/lepke.jpg"
                  />
                  <h3
                    class="font-bold text-xl sm:text-2xl mt-4"
                  >
                    LandingPage.products.2.title
                  </h3>
                  <p
                    class="mt-4"
                  >
                    LandingPage.products.2.description
                  </p>
                </a>
                <!--container-->
              </app-product-item>
              <!--container-->
            </div>
          </app-product-list>
        </section>
      </app-frame>
      <!--container-->
    </section>
  </app-page-layout>
  <!--container-->
  <!--container-->
</div>
`;
